File: modules/openlmis-web/src/main/java/org/openlmis/web/controller/vaccine/VaccineManufacturerController.java
/*
 * This program is part of the OpenLMIS logistics management information system platform software.
 * Copyright © 2013 VillageReach
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *  
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.
 * You should have received a copy of the GNU Affero General Public License along with this program.  If not, see http://www.gnu.org/licenses.  For additional information contact info@OpenLMIS.org. 
 */

package org.openlmis.web.controller.vaccine;

import org.openlmis.vaccine.domain.Manufacturer;
import org.openlmis.vaccine.domain.ManufacturerProduct;
import org.openlmis.vaccine.service.ManufacturerService;
import org.openlmis.web.controller.BaseController;
import org.openlmis.web.response.OpenLmisResponse;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.dao.DuplicateKeyException;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;

import javax.servlet.http.HttpServletRequest;

import java.util.List;

import static org.openlmis.web.response.OpenLmisResponse.success;
import static org.springframework.web.bind.annotation.RequestMethod.DELETE;
import static org.springframework.web.bind.annotation.RequestMethod.GET;
import static org.springframework.web.bind.annotation.RequestMethod.POST;

@Controller
@RequestMapping(value="/vaccine/manufacturer")
public class VaccineManufacturerController extends BaseController{


    @Autowired
    ManufacturerService manufacturerService;

    @RequestMapping(value = "/create", method = POST, headers = ACCEPT_JSON)
    // TODO: Add appropriate permission
    //  @PreAuthorize("@permissionEvaluator.hasPermission(principal,'MANAGE_VACCINE_MANUFACTURER')")
    public ResponseEntity insert(@RequestBody Manufacturer manufacturer, HttpServletRequest request) {

        manufacturer.setCreatedBy(loggedInUserId(request));
        manufacturer.setModifiedBy(loggedInUserId(request));

        ResponseEntity<OpenLmisResponse> response;

        try {
            manufacturerService.updateManufacturer(manufacturer);
        } catch(DuplicateKeyException exp){
            return OpenLmisResponse.error("There is record with the same vaccine quantification year already.", HttpStatus.BAD_REQUEST);
        }

        response = success("Vaccine manufacuted saved successfully");
        return response;
    }

    @RequestMapping(value = "/delete/{id}", method = DELETE, headers = ACCEPT_JSON)
    // TODO: Add appropriate permission
    //  @PreAuthorize("@permissionEvaluator.hasPermission(principal,'MANAGE_VACCINE_MANUFACTURER')")
    public ResponseEntity deleteVaccineManufacturer(@PathVariable(value="id") Long id){

        try{

            manufacturerService.deleteManufacturer(id);
        }
        catch(DataIntegrityViolationException exp){
            return OpenLmisResponse.error("There is a data dependent on the selected manufacturer", HttpStatus.BAD_REQUEST);
        }


        ResponseEntity<OpenLmisResponse> response;
        response = success("Vaccine manufacutrer deleted successfully");
        return response;
    }

    @RequestMapping(value = "/list", method = GET, headers = ACCEPT_JSON)
    public ResponseEntity getVaccineManufacturers() {

        return OpenLmisResponse.response("vaccineManufacturers", manufacturerService.getAll());
    }

    @RequestMapping(value = "/get/{id}", method = GET, headers = ACCEPT_JSON)
    public ResponseEntity getVaccineManufacturer(@PathVariable(value="id") Long id){
        return OpenLmisResponse.response("vaccineManufacturer", manufacturerService.getManufacturer(id));
    }

    /*** Manufacturer product Mapping *****/

    @RequestMapping(value = "/getManufacturerProducts/{id}", method = GET, headers = ACCEPT_JSON)
    public ResponseEntity getVaccineManufacturerProductMapping(@PathVariable(value="id") Long id){

        Manufacturer manufacturer = manufacturerService.getManufacturer(id);
        List<ManufacturerProduct> product = manufacturerService.getProductMapping(id);

        ResponseEntity<OpenLmisResponse> response;
        response = OpenLmisResponse.response("vaccineManufacturer", manufacturer);
        response.getBody().addData("productMapping", product);
        return response;
    }

    @RequestMapping(value = "/getProduct/{id}", method = GET, headers = ACCEPT_JSON)
    public ResponseEntity getProductMappingByMappingId(@PathVariable(value="id") Long id){
        return OpenLmisResponse.response("manufacturerProduct", manufacturerService.getProductMappingByMappingId(id));
    }

    @RequestMapping(value = "/deleteProduct/{id}", method = DELETE, headers = ACCEPT_JSON)
    public ResponseEntity deleteProductMapping(@PathVariable(value="id") Long id){

        manufacturerService.deleteProductMapping(id);

        ResponseEntity<OpenLmisResponse> response;
        response = success("Product mapping deleted successfully");
        return response;
    }

    @RequestMapping(value = "/createProduct", method = POST, headers = ACCEPT_JSON)
    public ResponseEntity createUpdate(@RequestBody ManufacturerProduct manufacturerProduct, HttpServletRequest request){

        manufacturerProduct.setCreatedBy(loggedInUserId(request));
        manufacturerProduct.setModifiedBy(loggedInUserId(request));

        ResponseEntity<OpenLmisResponse> response;

        try {
            manufacturerService.addUpdateProductMapping(manufacturerProduct);
        } catch(DuplicateKeyException exp) {
            return OpenLmisResponse.error("Duplicate key not allowed.", HttpStatus.BAD_REQUEST);
        }
        catch(DataIntegrityViolationException exp){
            return OpenLmisResponse.error("Invalid product code", HttpStatus.BAD_REQUEST);
        }

        response = success("Product mapping saved successfully");
        return response;
    }

}

File: modules/openlmis-web/src/main/webapp/public/js/vaccine/controller/vaccine-manufacturer-controller.js
/*
 * This program was produced for the U.S. Agency for International Development. It was prepared by the USAID | DELIVER PROJECT, Task Order 4. It is part of a project which utilizes code originally licensed under the terms of the Mozilla Public License (MPL) v2 and therefore is licensed under MPL v2 or later.
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the Mozilla Public License as published by the Mozilla Foundation, either version 2 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the Mozilla Public License for more details.
 *
 * You should have received a copy of the Mozilla Public License along with this program. If not, see http://www.mozilla.org/MPL/
 */


function VaccineManufacturerController($scope, $dialog, messageService, $routeParams, $location, GetVaccineManufacturerProductMapping, VaccineManufacturerList,Products,  VaccineManufacturerUpdate, GetVaccineManufacturer, DeleteVaccineManufacturer, DeleteManufacturerProductMapping) {

    $scope.vaccineManufacturers = [];


    // manufacturer list
   if (isUndefined($routeParams.id)){

        VaccineManufacturerList.get({}, function(data){
            $scope.vaccineManufacturers = data.vaccineManufacturers;
        });
    }
    // Manufacturer edit
   else {

       if ($routeParams.id > 0) {

           $scope.$parent.message = '';

           GetVaccineManufacturerProductMapping.get({
               id: $routeParams.id
           }, function (data) {
               $scope.manufacturer = data.vaccineManufacturer;
               $scope.manufacturer.registrationDate = data.vaccineManufacturer.registrationDateString;
               $scope.productMappings = data.productMapping;
           });
       }

       else{
           $scope.$parent.message = '';
       }
   }

    $scope.saveVaccineManufacturer = function(){
        if ($scope.vaccineManufacturerForm.$error.pattern || $scope.vaccineManufacturerForm.$error.required) {
            $scope.showError = "true";
            $scope.error = 'form.error';
            $scope.message = "";
            return;
        }

        var onSuccess = function(data){
            $scope.$parent.message = 'Your changes have been saved!';
            loadVaccineManufacturerList();
            $location.path('/manufacturer');
        };

        var onError = function(data){
            $scope.showError = true;
            $scope.error = data.data.error;
        };

        VaccineManufacturerUpdate.post($scope.manufacturer,onSuccess, onError);
    };

    $scope.cancelManufacturerForm = function(){
        $location.path('/manufacturer');
    };


    function loadVaccineManufacturerList(){
        VaccineManufacturerList.get({}, function(data){
            $scope.vaccineManufacturers = data.vaccineManufacturers;
        });
    }

    $scope.showRemoveVaccineManufacturerDialog = function (vaccineManufacturer) {
        $scope.selectedVaccineManufacturer = vaccineManufacturer;

        var options = {
            id: "removeVaccineManufacturerConfirmDialog",
            header: "Confirmation",
            body: "Please confirm that you want to delete vaccine Manufacturer <strong>"+ $scope.selectedVaccineManufacturer.name+"</strong>"
        };
        OpenLmisDialog.newDialog(options, $scope.removeVaccineManufacturerConfirm, $dialog, messageService);
    };

    $scope.removeVaccineManufacturerConfirm = function (result) {

        if (result) {
            DeleteVaccineManufacturer.delete({'id': $scope.selectedVaccineManufacturer.id},
             function(){
                loadVaccineManufacturerList();
                $scope.$parent.message = 'Vaccine Manufacturer deleted successfully!';
            }, function(data){
                $scope.$parent.message = "";
                $scope.error = data.data.error;
            });
        }
    };


    $scope.showRemoveManufacturerProductDialog = function (productMapping) {
        $scope.selectedProductMapping = productMapping;

        var options = {
            id: "removeManufacturerProductMappingConfirmDialog",
            header: "Confirmation",
            body: "Please confirm that you want to delete <strong>"+$scope.selectedProductMapping.productName+"</strong> from the mapping list"
        };
        OpenLmisDialog.newDialog(options, $scope.removeManufacturerProductMappingConfirm, $dialog, messageService);
    };

    $scope.removeManufacturerProductMappingConfirm = function (result) {

        if (result) {
            DeleteManufacturerProductMapping.delete({'id': $scope.selectedProductMapping.id}, function(){

                GetVaccineManufacturerProductMapping.get({
                    id: $routeParams.id
                }, function (data) {
                    $scope.manufacturer = data.vaccineManufacturer;
                    $scope.productMappings = data.productMapping;
                });

                $scope.$parent.message = 'Manufacturer product mapping deleted successfully!';
            });
        }
    };

}